hoot-core/src/main/cpp/hoot/core/conflate/splitter/CornerSplitter.cpp

```diff
@@ -0,0 +1,166 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * --------------------------------------------------------------------
+ *
+ * The following copyright notices are generated automatically. If you
+ * have a new notice to add, please use the format:
+ * " * @copyright Copyright ..."
+ * This will properly maintain the copyright information. DigitalGlobe
+ * copyrights will be updated automatically.
+ *
+ * @copyright Copyright (C) 2017 DigitalGlobe (http://www.digitalglobe.com/)
+ */
+
+#include "CornerSplitter.h"
+
+// Hoot
+#include <hoot/core/util/Factory.h>
+#include <hoot/core/elements/Way.h>
+#include <hoot/core/algorithms/WaySplitter.h>
+#include <hoot/core/algorithms/WayHeading.h>
+#include <hoot/core/index/OsmMapIndex.h>
+#include <hoot/core/schema/OsmSchema.h>
+#include <hoot/core/algorithms/linearreference/WayLocation.h>
+#include <hoot/core/OsmMap.h>
+
+// Qt
+#include <QDebug>
+
+using namespace std;
+
+namespace hoot
+{
+
+HOOT_FACTORY_REGISTER(OsmMapOperation, CornerSplitter)
+
+CornerSplitter::CornerSplitter()
+{
+
+}
+
+CornerSplitter::CornerSplitter(boost::shared_ptr<OsmMap> map)
+{
+  _map = map;
+}
+
+void CornerSplitter::splitCorners(boost::shared_ptr<OsmMap> map)
+{
+  CornerSplitter splitter(map);
+  return splitter.splitCorners();
+}
+
+void CornerSplitter::splitCorners()
+{
+  // Get a list of ways in the map
+  for (WayMap::const_iterator it = _map->getWays().begin(); it != _map->getWays().end(); ++it)
+  {
+    _todoWays.push_back(it->first);
+  }
+
+  // Traverse each way, looking for corners to split
+  // Splitting a way will usually result in adding a new way to _todoWays
+  for (size_t i = 0; i < _todoWays.size(); i++)
+  {
+    ConstWayPtr pWay = _map->getWay(_todoWays[i]);
+    size_t nodeCount = pWay->getNodeCount();
+
+    // If the way has just two nodes, there are no corners
+    if (nodeCount > 2)
+    {
+      for (size_t nodeIdx = 1; nodeIdx < nodeCount-1; nodeIdx++)
+      {
+        WayLocation prev(_map, pWay, nodeIdx-1, 0.0);
+        WayLocation current(_map, pWay, nodeIdx, 0.0);
+        WayLocation next(_map, pWay, nodeIdx+1, 0.0);
+
+        // Calculate headings
+        double h1 = atan2(current.getCoordinate().y - prev.getCoordinate().y,
+                          current.getCoordinate().x - prev.getCoordinate().x);
+        double h2 = atan2(next.getCoordinate().y - current.getCoordinate().y,
+                          next.getCoordinate().x - current.getCoordinate().x);
+
+        double threshold = toRadians(45.0);
+        double delta = fabs(h2-h1);
+
+
+        // If we make enough of a turn, split the way
+        if (delta > threshold)
+        {
+          _splitWay(pWay->getId(), nodeIdx, pWay->getNodeId(nodeIdx));
+        }
+      }
+    }
+  }
+}
+
+void CornerSplitter::_splitWay(long wayId, long nodeIdx, long nodeId)
+{
```

Comment:
There is already a `WaySplitter` class that does a very similar thing (if not the same).  It might be easier to call the static function:
```
static std::vector< WayPtr > split(const OsmMapPtr& map, WayPtr a,
    WayLocation& splitPoint);
```
to make the splits.  The class might even provide more functionality than just that.  If it isn't exactly what you want, could functionality be added to that class to consolidate the way splitting into one class?

Reply:
I will definitely use the WaySplitter class. It shouldn't be a problem!

Reply:
Just kidding. That _splitWay function makes use of the WaySplitter. The bulk of the function is logging, and processing the splits that are generated. I don't think split processing (deciding what to do with the resulting ways) belongs in the WaySplitter class - so I'll leave this function as-is.